Simplified/Reader2/Bookmarks/NYPLAnnotations.swift

```diff
@@ -590,7 +465,116 @@ import R2Shared
+    if let error = error as NSError? {
+      Log.error(#file, "Request Error Code: \(error.code). Description: \(error.localizedDescription)")
+      return nil
+    }
+
+    guard let data = data,
+      let jsonObject = try? JSONSerialization.jsonObject(with: data, options: []),
+      let json = jsonObject as? [String: Any] else {
+        Log.error(#file, "Response from annotation server could not be serialized.")
+        return nil
+    }
+
+    guard let first = json["first"] as? [String: Any],
+      let items = first["items"] as? [[String: Any]] else {
+        Log.error(#file, "Missing required key from Annotations response, or no items exist.")
+        return nil
+    }
+
+    let bookmarks = items.compactMap {
+      NYPLBookmarkFactory.make(fromServerAnnotation: $0,
+                               annotationType: motivation,
+                               bookID: bookID,
+                               publication: publication)
+    }
+
+    return bookmarks
+  }
+
+  /// Serializes the `parameters` into JSON and POSTs them to the server.
+  private class func postAnnotation(
+    forBook bookID: String,
+    withParameters parameters: [String: Any],
+    timeout: TimeInterval = NYPLDefaultRequestTimeout,
+    queueOffline: Bool,
+    _ completionHandler: @escaping (_ success: Bool, _ annotationID: String?) -> ()) {
+
+    guard syncIsPossibleAndPermitted() else {
+      Log.debug(#file, "Account does not support sync or sync is disabled.")
+      completionHandler(false, nil)
+      return
+    }
+
+    guard let annotationsURL = NYPLAnnotations.annotationsURL else {
+      Log.error(#file, "Annotations URL was nil while attempting to post")
+      completionHandler(false, nil)
+      return
+    }
+
+    guard let jsonData = makeSubmissionData(fromRepresentation: parameters) else {
+      Log.error(#file, "Network request abandoned. Could not create JSON from given parameters.")
+      completionHandler(false, nil)
+      return
+    }
+
+    var request = URLRequest(url: annotationsURL)
+    request.httpMethod = "POST"
+    request.httpBody = jsonData
+    setDefaultAnnotationHeaders(forRequest: &request)
+    request.timeoutInterval = timeout
+
+    let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
+      if let error = error as NSError? {
+        Log.error(#file, "Annotation POST error (nsCode: \(error.code) Description: \(error.localizedDescription))")
+        if (NetworkQueue.StatusCodes.contains(error.code)) && (queueOffline == true) {
+          self.addToOfflineQueue(bookID, annotationsURL, parameters)
+        }
+        completionHandler(false, nil)
+        return
+      }
+      guard let statusCode = (response as? HTTPURLResponse)?.statusCode else {
+        Log.error(#file, "Annotation POST error: No response received from server")
+        completionHandler(false, nil)
+        return
+      }
+
+      if statusCode == 200 {
```

Comment:
Is there a chance that the server might return 201 upon success at some point? I think we should consider any status code within the range 200...299 as success.

Reply:
I'm not sure -- this is just code I moved from one place to another, and it's out of scope of this ticket, so I was just going to leave it as is. Really what should be done here is a bigger refactor for the whole request execution, i.e. going through out NYPLNetworkExecutor layer. That way all of this error handling will be done for free (including proper status code detection)